Add dependencies to makefile install target

4ydan opened the pull request:
#679

## Basics

<!--
These points need to be fulfilled for every PR.
-->

- [ ] I added a line to [/doc/CHANGELOG.md](/doc/CHANGELOG.md)
- [ ] The PR is rebased with current master.
- [ ] Details of what you changed are in commit messages.
- [ ] References to issues, e.g. `close #X`, are in the commit messages.
- [ ] The buildserver is happy.

<!--
If you have any troubles fulfilling these criteria, please write about the trouble as comment in the PR.
We will help you, but we cannot accept PRs that do not fulfill the basics.
-->

## Checklist

<!--
For documentation fixes, spell checking, and similar none of these points below need to be checked.
Otherwise please check these points when getting a PR done:
-->

- [ ] I have installed and I am using [pre-commit hooks](../doc/contrib/README.md#Hooks)
- [ ] I fully described what my PR does in the documentation
      (not in the PR description)
- [ ] I fixed all affected documentation
- [ ] I fixed all affected decisions
- [ ] I added unit tests for my code
- [ ] I added code comments, logging, and assertions as appropriate
- [ ] I translated all strings visible to the user
- [ ] I mentioned [every code or binary](https://github.com/ElektraInitiative/PermaplanT/blob/master/.reuse/dep5) not directly written or done by me in [reuse syntax](https://reuse.software/)
- [ ] I created left-over issues for things that are still to be done
- [ ] Code is conforming to [our Architecture](/doc/architecture)
- [ ] Code is conforming to [our Guidelines](/doc/guidelines)
      (exceptions are documented)
- [ ] Code is consistent to [our Design Decisions](/doc/decisions)

## Review

<!--
Reviewers can copy&check the following to their review.
Also the checklist above can be used.
But also the PR creator should check these points when getting a PR done:
-->

- [ ] I've tested the code
- [ ] I've read through the whole code
- [ ] Examples are well chosen and understandable

markus2330 commented on Makefile:
```diff
@@ -106,40 +111,31 @@ psql-r:  ## Remote connect to postgis, uses $POSTGRES_USER and $POSTGRES_DB.
-clean-scraper:  ## Remove /scraper/node_modules .
-	cd scraper && rm -rf node_modules
+.PHONY: install
+install:  $(FAKE_FILE)  node_modules  ## Install ALL dependencies within the source repo.
```
> package.json somehow doesn't work. The other install with creating a .install file work fine.

What seems to be the problem? I actually thought that must be trivial?

Following you tried:

```suggestion
install:  $(FAKE_FILE)  frontend/package.json  ## Install ALL dependencies within the source repo.
```

Did you already consider to also have a frontend Makefile? In general we should look out to keep everything consistent. Can't we name `node_modules` to something like `install-node-modules`?

> Also simplified it a bit.

Always great :heart:

markus2330 commented on README.md:
```diff
@@ -21,19 +21,9 @@ Which will open [/doc/architecture/README.md](/doc/architecture/README.md) as fi
-Type `make help` to see all commands
-
-**SOME OF THESE COMMANDS REQUIRE ENV VARIABLES, LOOK AT [backend env variables](./doc/backend/01setup.md) or [frontend env variables](./frontend/README.md)**
+**Some of the commands require env variables to be set. Have a look at [backend env variables](./doc/backend/01setup.md) and [frontend env variables](./frontend/README.md)**
```
Please also add scraper.

```suggestion
**Some of the commands require env variables to be set or `.env` being present.
Have a look at [backend env variables](./doc/backend/01setup.md) and [frontend env variables](./frontend/README.md)**
```

markus2330 reviewed (changes requested):
There are also a few unnecessary deps:
E.g. build-backend does not need to depend on generate-api-types (only build-frontend needs it).

4ydan commented:
I am thinking of some sort of database-reset + scraper insert in one target. What do you think?

markus2330 commented:
> I am thinking of some sort of database-reset + scraper-insert in one target. What do you think?

Yes, great idea! I use this often, too.

> Maybe we could include the scraper-insert into database-reset?

Yes, reset-database sounds like it should import everything again.

I would swap verb+noun, as most other actions have the verb first.

4ydan commented:
Ups I pressed merge...

4ydan commented:
This is skipping now npm install when the node modules is newer than package.json
and
skip installing rustfmt and mdbook dependencies depending on a .install fake file.

markus2330 commented on Makefile:
```diff
@@ -1,162 +1,188 @@
 .PHONY: install-e2e
-install-e2e:
-	cd e2e && ./install.sh
+install-e2e:  ## Install e2e dependencies within this repo
```
I would also include that in install (so that targets without specifier always do everything) if it does nothing if already done.

markus2330 reviewed (approved):
Fascinating work. Needs rebase. Maybe we should make `install` do more, to have more consistency in our targets.

4ydan commented:
> Btw. the install-e2e used to need a root password. I would prefer if everything could be installed within the PermaplanT folder and definitely everything as user. Is this possible?

Maybe should have done it with venv from the start. Hopefully it works now.

4ydan commented:
Ok every install checks something to see if an install is necessary.

- backend checks the fakefile
- frontend checks mtime node_modules > package.json
- e2e checks mtime venv > (requirements.txt || install.sh)

4ydan commented:
> But with pre commit -a https://github.com/ElektraInitiative/PermaplanT/pull/736 it should actually show up?

The `mdbook test` stage, should have also revealed that error.

badnames reviewed (approved):
Looks good to me.

temmey commented on README.md:
```diff
@@ -21,19 +21,9 @@ Which will open [/doc/architecture/README.md](/doc/architecture/README.md) as fi
-Type `make help` to see all commands
-
-**SOME OF THESE COMMANDS REQUIRE ENV VARIABLES, LOOK AT [backend env variables](./doc/backend/01setup.md) or [frontend env variables](./frontend/README.md)**
+**Some of the commands require env variables to be set. Have a look at [backend env variables](./doc/backend/01setup.md), [frontend env variables](./frontend/README.md) and [scraper env variables](./scraper/README.md)**
```
```suggestion
**Some of the commands require env variables to be set. 
Have a look at [backend env variables](./doc/backend/01setup.md), [frontend env variables](./frontend/README.md) and [scraper env variables](./scraper/README.md)**
```

4ydan commented on e2e/install.sh:
```diff
@@ -1,3 +1,8 @@
 #!/bin/bash
+apt update
```
That is a very good point.

markus2330 commented on e2e/e2e.sh:
```diff
@@ -1,4 +1,5 @@
 #!/bin/bash
 
 # Run the pytest command
+source venv/bin/activate
```
This has strict requirements on the current working directory. Either document it or use `$0`  to locate it relative to the command itself.

markus2330 reviewed (changes requested):
Thx for the update! Better avoid unrelated changes in the same PR, I think the Makefile part is okay (and would be amazing to have this merged soon) but for e2e I think we can do better (imho it is very surprising that a package manager gets executed when running `make`).

markus2330 commented on Makefile:
```diff
@@ -1,162 +1,217 @@
-build-storybook: install generate-api-types  ## Build storybook.
-	cd frontend && npm install && npm run build-storybook
+build-storybook: generate-api-types  ## Build Storybook
+	@cd frontend && npm install && npm run build-storybook
```
Here is a bug, to make the storybook complete, also npm run doc is needed before:

```suggestion
	@cd frontend && npm install && npm run doc && npm run build-storybook
```

markus2330 reviewed (approved):
Great work, let get this done! :rocket

4ydan commented on Makefile:
```diff
@@ -1,162 +1,217 @@
+test-mdbook:  build-mdbook  ## Build & Test Mdbook
+	@mdbook test
+
+# Test all makefile targets except the ones with "run-".
+.PHONY: test-makefile
+test-makefile: $(MAKEFILE_LIST)
```
This is just for testing the makefile itself, should probably remove the phony.

markus2330 replied:
It basically executes everything but what starts with `run-`? What it does additionally, so that we can call it test? (I only refer about the documentation, the target itself is okay if you want/need it :wink:)

4ydan commented:
There is just something odd here now.
We created this PR to not perform make pre-requisites all the time.
But now if we want to use make targets in CI, some pre-requisites are redundant there, since packages are already installed in the docker image or the docker image doesnt have the dependencies to perform the installation.
Example:
mdbook image already has everything thats in `install-mdbook`
node image doesnt have typeshare to perform `generate-api-types`

markus2330 commented:
> But now if we want to use make targets in CI, some pre-requisites are redundant there, since packages are already installed in the docker image or the docker image doesnt have the dependencies to perform the installation.

You simply could do `touch .install` in CI.

> If we decide to remove the pre-requisites,

Currently I think that the only needed change is that we require `python3` to be present before `make install` is executed. The other stuff (mdbook etc.) can stay as it is (install it during `make install`)

We somehow need to clarify the pre-requisites for beginners. But I assume for doc/development_setup.md it does not matter as python3 is installed in Docker?

4ydan commented:
Using makefile2graps, I changed the whole structure a bit, making use everywhere of requisites and removed some duplications
![image](https://github.com/ElektraInitiative/PermaplanT/assets/53004139/8a436057-576e-47a4-82b5-cebc8b231564)

Does this graph look reasonable?

markus2330 commented:
Yes, looks very nice. Is for install-frontend and install-mdbook now also a way todetect if they are up-to-date or only for install-backend?

4ydan commented:
For `install-mdbook` not yet, but for all other installs there is (pre-commit, backend, e2e, frontend, types)

markus2330 commented:
The PR needs a rebase. Would be great to get this soon.

markus2330 commented:
CI not yet finished?

markus2330 commented:
@4ydan is this ready to merge?

4ydan commented:
Not yet, I want to look over everything again, once I have more time.

4ydan commented:
Can merge this.

markus2330 commented:
Great job to have this finally done! :heart: :rocket: 

Btw. IIRC it was actually planned as last issue if you happen to have additional time :wink:

4ydan commented:
Yes, it was just not much to do here, thats why I decided to finish it up quick.
I worked on the e2e tests, but I can barely progress.

markus2330 commented:
Can you maybe ping the issues which hold you back so that they can be done first?

An issue for 0.3.4 release coordination might be helpful, too.
